libvips/foreign/spngsave.c

```diff
@@ -0,0 +1,848 @@
+{
+	VipsObjectClass *class = VIPS_OBJECT_GET_CLASS( spng );
+	VipsForeignSave *save = (VipsForeignSave *) spng;
+
+	int error;
+	struct spng_ihdr ihdr;
+	struct spng_phys phys;
+	struct spng_plte plte = { 0 };
+	struct spng_trns trns = { 0 };
+	int fmt;
+	enum spng_encode_flags encode_flags;
+
+	spng->ctx = spng_ctx_new( SPNG_CTX_ENCODER );
+
+	if( (error = spng_set_png_stream( spng->ctx, 
+		vips_foreign_save_spng_write_fn, spng )) ) {
+		vips_error( class->nickname, "%s", spng_strerror( error ) ); 
+		return( -1 );
+	}
+
+#ifdef HAVE_IMAGEQUANT
+	if( spng->palette ) {
+		VipsImage *im_index;
+		VipsImage *im_palette;
+		int palette_count;
+		int i;
+
+		if( vips__quantise_image( in, &im_index, &im_palette, 
+			1 << spng->bitdepth, 
+			spng->Q, 
+			spng->dither, 
+			spng->effort, 
+			FALSE ) )
+			return( -1 );
+
+		/* PNG is 8-bit index only.
+		 */
+		palette_count = im_palette->Xsize;
+		g_assert( palette_count <= 256 );
+
+		for( i = 0; i < palette_count; i++ ) {
+			VipsPel *p = (VipsPel *) 
+				VIPS_IMAGE_ADDR( im_palette, i, 0 );
+
+			if( p[3] == 255 ) {
+				struct spng_plte_entry *entry = 
+					&plte.entries[plte.n_entries];
+
+				entry->red = p[0];
+				entry->green = p[1];
+				entry->blue = p[2];
+				plte.n_entries += 1;
+			}
+			else {
+				trns.type3_alpha[trns.n_type3_entries] = p[3];
+				trns.n_type3_entries += 1;
+			}
+		}
+
+#ifdef DEBUG
+		printf( "attaching %d entry palette\n", plte.n_entries );
+		if( trns.n_type3_entries )
+			printf( "attaching %d transparency values\n", 
+			     trns.n_type3_entries );
+#endif /*DEBUG*/
+
+		VIPS_UNREF( im_palette );
+
+		in = spng->memory = im_index;
+	}
+#endif /*HAVE_IMAGEQUANT*/
+
+	ihdr.width = in->Xsize;
+	ihdr.height = in->Ysize;
+	ihdr.bit_depth = spng->bitdepth;
+	/* FIXME ... spng 0.7 does not yet implement 1/2/4 bit packing, so
+	 * you'll see a bad result.
+	 */
+	if( spng->bitdepth < 8 ) 
+		g_warning( "spngsave with bitdepth < 8 not yet supported" );
+
+	switch( in->Bands ) {
+	case 1:
+		if( spng->palette )
+			ihdr.color_type = SPNG_COLOR_TYPE_INDEXED; 
+		else
+			ihdr.color_type = SPNG_COLOR_TYPE_GRAYSCALE;
+		break;
+
+	case 2:
+		ihdr.color_type = SPNG_COLOR_TYPE_GRAYSCALE_ALPHA; 
+		break;
+
+	case 3:
+		ihdr.color_type = SPNG_COLOR_TYPE_TRUECOLOR;
+		break;
+
+	case 4: 
+		ihdr.color_type = SPNG_COLOR_TYPE_GRAYSCALE_ALPHA; 
+		break;
+
+	default:
+		vips_error( class->nickname, "%s", _( "bad bands" ) );
+		return( -1 );
+	}
+
+	ihdr.compression_method = 0;
+	ihdr.filter_method = 0;
+	ihdr.interlace_method = spng->interlace ? 1 : 0;
+	if( (error = spng_set_ihdr( spng->ctx, &ihdr )) ) {
+		vips_error( class->nickname, "%s", spng_strerror( error ) ); 
+		return( -1 );
+	}
+
+	spng_set_option( spng->ctx, 
+		SPNG_IMG_COMPRESSION_LEVEL, spng->compression );
+	spng_set_option( spng->ctx, 
+		SPNG_TEXT_COMPRESSION_LEVEL, spng->compression );
+	spng_set_option( spng->ctx, 
+		SPNG_FILTER_CHOICE, spng->filter );
+
```

Comment:
I had a quick experiment with changing libspng default [`Z_FILTERED`](https://github.com/randy408/libspng/blob/8c74bfe96bcb17123d8e83de1edeffe443b75f24/spng/spng.c#L4869) strategy to `Z_RLE`, since I remembered [this discussion](https://github.com/lovell/sharp/commit/3917efdebd91ddaf3878a276a6c7b79edce2d2fd#r41720871) with Lovell.
```suggestion

	/* Use Z_RLE strategy by default as that is specifically suitable for PNG.
	 */
	spng_set_option( spng->ctx,
		SPNG_IMG_COMPRESSION_STRATEGY, 3/*Z_RLE*/ );

```

After testing this with the `large_rgb8.png` image from https://github.com/libspng/benchmark_images/, I see this:
```bash
$ curl -LO https://github.com/libspng/benchmark_images/raw/master/large_rgb8.png
$ LD_PRELOAD="/usr/lib64/libz-ng.so.2.0.5" /usr/bin/time -f %M:%e vips copy large_rgb8.png z_filterd.png
32516:3.04
# After applying the above Z_RLE patch
$ LD_PRELOAD="/usr/lib64/libz-ng.so.2.0.5" /usr/bin/time -f %M:%e vips copy large_rgb8.png z_rle.png
32532:0.74
$ ls -al z_*.png
-rw-r--r--. 1 kleisauke kleisauke 13160679 Nov 26 15:01 z_filterd.png
-rw-r--r--. 1 kleisauke kleisauke 13181100 Nov 26 15:02 z_rle.png
```

/me falls off chair.

Reply:
wat

Reply:
I tried with a 10k x 10k RGB image:

```
$ time vips copy wtc.jpg x.png
real	0m27.542s
user	0m27.767s
sys	0m0.256s
$ time vips copy wtc.jpg x-rle.png
real	0m13.399s
user	0m13.718s
sys	0m0.188s
$ ls -l x.png x-rle.png 
-rw-r--r-- 1 john john 135800382 Nov 26 16:45 x.png
-rw-r--r-- 1 john john 146518124 Nov 26 16:46 x-rle.png
```

deflate compression defaults to 6. I tried turning that down to try to match the compression speed:

```
$ time vips copy wtc.jpg x.png[compression=1]
real	0m14.923s
user	0m15.166s
sys	0m0.265s
$ ls -l x.png
-rw-r--r-- 1 john john 137881560 Nov 26 16:50 x.png
```

So maybe deflate level 1 might be another default to consider? Better compression than RLE, and almost as fast.

Reply:
I'd expect RLE to suit images with long runs of the same pixel value, e.g. icons, line art and perhaps palette-based images, which is typically what PNG is used for on the web, plus I guess it might also use a bit less memory. Interesting that the examples here are photographic in nature, which also seems to be a partial conclusion of http://optipng.sourceforge.net/pngtech/z_rle.html